File: frontend/src/editor/commands.ts
// Editor action layer.
//
// A single, typed vocabulary of edit operations. Both the toolbar/mode UI and
// the AI agent describe an edit as an `EditorCommand`; `applyCommand` turns it
// into a new TimelineData using the pure ops in timeline.ts. Keeping every
// mutation funnelled through here means the agent and the buttons share exactly
// one code path, and each command is trivially undoable (it's just a producer
// passed to `editor.commit`).
//
// `COMMAND_REGISTRY` is the machine-readable catalogue of everything the editor
// can do — its `params` schema is what an AI is handed so it can call any
// operation, including ones that have no button yet ("AI-only"). UI buttons/
// inspectors are just convenient front-ends over the same commands.

import type {
  ColorGrade,
  MaskRect,
  TimelineData,
  TimelineElement,
  Transform,
} from '../api/client'
import {
  addClip,
  addTrack,
  deleteClip,
  duplicateClips,
  findClip,
  linkClips,
  makeTextClip,
  moveClipGroup,
  moveTrack,
  removeTrack,
  rippleDelete,
  setTrackFlags,
  splitClip,
  trimEnd,
  trimStart,
  unlinkClips,
  updateClip,
} from './timeline'
import { NEUTRAL_TRANSFORM } from './effects'

export type EditorCommand =
  // ---- Timeline structure ----
  | { type: 'split'; clipId: string; at: number }
  | { type: 'delete'; clipIds: string[] }
  | { type: 'rippleDelete'; clipIds: string[] }
  | { type: 'duplicate'; clipIds: string[] }
  | { type: 'moveClip'; clipId: string; start: number; trackId?: string }
  | { type: 'trimStart'; clipId: string; at: number }
  | { type: 'trimEnd'; clipId: string; at: number; sourceMax?: number }
  // ---- Generic props ----
  | { type: 'setProps'; clipIds: string[]; patch: Partial<TimelineElement> }
  // ---- Audio ----
  | { type: 'setVolume'; clipIds: string[]; volume: number }
  | { type: 'setAudioFade'; clipIds: string[]; audioFadeIn?: number; audioFadeOut?: number }
  // ---- Opacity / fades ----
  | { type: 'setOpacity'; clipIds: string[]; opacity: number }
  | { type: 'setFade'; clipIds: string[]; fadeIn?: number; fadeOut?: number }
  // ---- Speed ----
  | { type: 'setSpeed'; clipIds: string[]; speed: number }
  // ---- Color ----
  | { type: 'setColor'; clipIds: string[]; color: Partial<ColorGrade> }
  // ---- Transform ----
  | { type: 'setScale'; clipIds: string[]; scale: number }
  | { type: 'setRotation'; clipIds: string[]; degrees: number }
  | { type: 'rotateBy'; clipIds: string[]; degrees: number }
  | { type: 'nudge'; clipIds: string[]; dx: number; dy: number }
  | { type: 'resetTransform'; clipIds: string[] }
  | { type: 'flipH'; clipIds: string[] }
  | { type: 'flipV'; clipIds: string[] }
  // ---- Crop (mask) ----
  | { type: 'setCrop'; clipIds: string[]; crop: MaskRect | null }
  // ---- Linking (magnet) ----
  | { type: 'link'; clipIds: string[] }
  | { type: 'unlink'; clipIds: string[] }
  // ---- Text ----
  | { type: 'addText'; at: number; trackId?: string; text?: string; x?: number; y?: number }
  // ---- Tracks ----
  | { type: 'addTrack'; kind: 'video' | 'audio' | 'text' }
  | { type: 'removeTrack'; trackId: string }
  | { type: 'moveTrack'; trackId: string; dir: -1 | 1 }
  | { type: 'setTrackHidden'; trackId: string; hidden: boolean }
  | { type: 'setTrackLocked'; trackId: string; locked: boolean }

// Merge a partial transform onto a clip's current transform.
function mergeTransform(el: TimelineElement, patch: Partial<Transform>): Transform {
  return { ...NEUTRAL_TRANSFORM, ...(el.transform ?? {}), ...patch }
}

// Human-readable label for a command (used by the agent transcript / history).
export function describeCommand(cmd: EditorCommand): string {
  switch (cmd.type) {
    case 'split':
      return `Split clip at ${cmd.at.toFixed(2)}s`
    case 'delete':
      return `Delete ${cmd.clipIds.length} clip(s)`
    case 'rippleDelete':
      return `Ripple-delete ${cmd.clipIds.length} clip(s)`
    case 'duplicate':
      return `Duplicate ${cmd.clipIds.length} clip(s)`
    case 'moveClip':
      return `Move clip to ${cmd.start.toFixed(2)}s`
    case 'trimStart':
      return `Trim clip start to ${cmd.at.toFixed(2)}s`
    case 'trimEnd':
      return `Trim clip end to ${cmd.at.toFixed(2)}s`
    case 'setProps':
      return `Update ${cmd.clipIds.length} clip(s)`
    case 'setVolume':
      return `Set volume ${Math.round(cmd.volume * 100)}%`
    case 'setAudioFade':
      return `Set audio fade`
    case 'setOpacity':
      return `Set opacity ${Math.round(cmd.opacity * 100)}%`
    case 'setFade':
      return `Set fade`
    case 'setSpeed':
      return `Set speed ${cmd.speed}×`
    case 'setColor':
      return `Color grade`
    case 'setScale':
      return `Scale ${cmd.scale.toFixed(2)}×`
    case 'setRotation':
      return `Rotate to ${cmd.degrees}°`
    case 'rotateBy':
      return `Rotate ${cmd.degrees > 0 ? '+' : ''}${cmd.degrees}°`
    case 'nudge':
      return `Nudge clip`
    case 'resetTransform':
      return `Reset transform`
    case 'flipH':
      return `Flip horizontal`
    case 'flipV':
      return `Flip vertical`
    case 'setCrop':
      return cmd.crop ? `Crop` : `Clear crop`
    case 'link':
      return `Link ${cmd.clipIds.length} clip(s)`
    case 'unlink':
      return `Unlink clip(s)`
    case 'addText':
      return `Add text`
    case 'addTrack':
      return `Add ${cmd.kind} track`
    case 'removeTrack':
      return `Remove track`
    case 'moveTrack':
      return `Move track ${cmd.dir < 0 ? 'up' : 'down'}`
    case 'setTrackHidden':
      return cmd.hidden ? `Hide track` : `Show track`
    case 'setTrackLocked':
      return cmd.locked ? `Lock track` : `Unlock track`
  }
}

function updateEach(
  data: TimelineData,
  clipIds: string[],
  patch: (el: TimelineElement) => Partial<TimelineElement>,
): TimelineData {
  let d = data
  for (const id of clipIds) {
    const found = findClip(d, id)
    if (found) d = updateClip(d, id, patch(found.el))
  }
  return d
}

export function applyCommand(data: TimelineData, cmd: EditorCommand): TimelineData {
  switch (cmd.type) {
    case 'split':
      return splitClip(data, cmd.clipId, cmd.at)

    case 'delete': {
      let d = data
      for (const id of cmd.clipIds) d = deleteClip(d, id)
      return d
    }

    case 'rippleDelete':
      return rippleDelete(data, cmd.clipIds)

    case 'duplicate':
      return duplicateClips(data, cmd.clipIds)

    case 'moveClip':
      return moveClipGroup(data, cmd.clipId, cmd.start, cmd.trackId)

    case 'trimStart':
      return trimStart(data, cmd.clipId, cmd.at)

    case 'trimEnd':
      return trimEnd(data, cmd.clipId, cmd.at, cmd.sourceMax)

    case 'setProps': {
      let d = data
      for (const id of cmd.clipIds) d = updateClip(d, id, cmd.patch)
      return d
    }

    case 'setVolume':
      return updateEach(data, cmd.clipIds, () => ({ volume: clamp01(cmd.volume) }))

    case 'setAudioFade':
      return updateEach(data, cmd.clipIds, () => ({
        ...(cmd.audioFadeIn != null ? { audioFadeIn: Math.max(0, cmd.audioFadeIn) } : {}),
        ...(cmd.audioFadeOut != null ? { audioFadeOut: Math.max(0, cmd.audioFadeOut) } : {}),
      }))

    case 'setOpacity':
      return updateEach(data, cmd.clipIds, () => ({ opacity: clamp01(cmd.opacity) }))

    case 'setFade':
      return updateEach(data, cmd.clipIds, () => ({
        ...(cmd.fadeIn != null ? { fadeIn: Math.max(0, cmd.fadeIn) } : {}),
        ...(cmd.fadeOut != null ? { fadeOut: Math.max(0, cmd.fadeOut) } : {}),
      }))

    case 'setSpeed':
      return updateEach(data, cmd.clipIds, () => ({ speed: Math.max(0.1, cmd.speed) }))

    case 'setColor':
      return updateEach(data, cmd.clipIds, (el) => ({
        color: {
          brightness: cmd.color.brightness ?? el.color?.brightness ?? 1,
          contrast: cmd.color.contrast ?? el.color?.contrast ?? 1,
          saturation: cmd.color.saturation ?? el.color?.saturation ?? 1,
        },
      }))

    case 'setScale':
      return updateEach(data, cmd.clipIds, (el) => ({
        transform: mergeTransform(el, { scale: Math.max(0.05, cmd.scale) }),
      }))

    case 'setRotation':
      return updateEach(data, cmd.clipIds, (el) => ({
        transform: mergeTransform(el, { rotation: cmd.degrees }),
      }))

    case 'rotateBy':
      return updateEach(data, cmd.clipIds, (el) => ({
        transform: mergeTransform(el, {
          rotation: (el.transform?.rotation ?? 0) + cmd.degrees,
        }),
      }))

    case 'nudge':
      return updateEach(data, cmd.clipIds, (el) => ({
        transform: mergeTransform(el, {
          x: (el.transform?.x ?? 0) + cmd.dx,
          y: (el.transform?.y ?? 0) + cmd.dy,
        }),
      }))

    case 'resetTransform':
      return updateEach(data, cmd.clipIds, () => ({ transform: { ...NEUTRAL_TRANSFORM } }))

    case 'flipH':
      return updateEach(data, cmd.clipIds, (el) => ({ flipH: !el.flipH }))

    case 'flipV':
      return updateEach(data, cmd.clipIds, (el) => ({ flipV: !el.flipV }))

    case 'setCrop':
      return updateEach(data, cmd.clipIds, () => ({ crop: cmd.crop }))

    case 'link':
      return linkClips(data, cmd.clipIds)

    case 'unlink':
      return unlinkClips(data, cmd.clipIds)

    case 'addText': {
      const trackId = cmd.trackId ?? data.tracks.find((t) => t.kind === 'text')?.id
      if (!trackId) return data
      const clip = makeTextClip(cmd.text ?? 'New text', cmd.at)
      if (cmd.x != null || cmd.y != null) {
        clip.transform = { ...NEUTRAL_TRANSFORM, x: cmd.x ?? 0, y: cmd.y ?? 0 }
      }
      return addClip(data, trackId, clip)
    }

    case 'addTrack':
      return addTrack(data, cmd.kind, 0)

    case 'removeTrack':
      return removeTrack(data, cmd.trackId)

    case 'moveTrack':
      return moveTrack(data, cmd.trackId, cmd.dir)

    case 'setTrackHidden':
      return setTrackFlags(data, cmd.trackId, { hidden: cmd.hidden })

    case 'setTrackLocked':
      return setTrackFlags(data, cmd.trackId, { locked: cmd.locked })
  }
}

function clamp01(v: number): number {
  return Math.max(0, Math.min(1, v))
}

// ---- Machine-readable command catalogue (for AI tool-calling) --------------

export interface CommandParam {
  name: string
  type: 'number' | 'string' | 'boolean' | 'clipIds' | 'trackId' | 'rect'
  optional?: boolean
  desc: string
}

export interface CommandSpec {
  id: EditorCommand['type']
  label: string
  desc: string
  params: CommandParam[]
  // Whether the editor exposes a dedicated button/control for it. AI-only
  // commands (false) are fully functional but simply have no UI yet.
  hasUI: boolean
}

const CLIP_IDS: CommandParam = {
  name: 'clipIds',
  type: 'clipIds',
  desc: 'Ids of the clips to act on',
}

export const COMMAND_REGISTRY: CommandSpec[] = [
  { id: 'split', label: 'Split', desc: 'Cut a clip in two at a time', hasUI: true, params: [{ name: 'clipId', type: 'string', desc: 'Clip id' }, { name: 'at', type: 'number', desc: 'Timeline seconds' }] },
  { id: 'delete', label: 'Delete', desc: 'Remove clips', hasUI: true, params: [CLIP_IDS] },
  { id: 'rippleDelete', label: 'Ripple delete', desc: 'Remove clips and close the gap', hasUI: true, params: [CLIP_IDS] },
  { id: 'duplicate', label: 'Duplicate', desc: 'Copy clips', hasUI: true, params: [CLIP_IDS] },
  { id: 'moveClip', label: 'Move', desc: 'Move a clip (and its link group) to a start time, optionally to another track', hasUI: true, params: [{ name: 'clipId', type: 'string', desc: 'Clip id' }, { name: 'start', type: 'number', desc: 'New start (s)' }, { name: 'trackId', type: 'trackId', optional: true, desc: 'Target track' }] },
  { id: 'trimStart', label: 'Trim start', desc: 'Set a clip in-point', hasUI: true, params: [{ name: 'clipId', type: 'string', desc: 'Clip id' }, { name: 'at', type: 'number', desc: 'New start (s)' }] },
  { id: 'trimEnd', label: 'Trim end', desc: 'Set a clip out-point', hasUI: true, params: [{ name: 'clipId', type: 'string', desc: 'Clip id' }, { name: 'at', type: 'number', desc: 'New end (s)' }] },
  { id: 'setProps', label: 'Set properties', desc: 'Patch arbitrary clip fields', hasUI: false, params: [CLIP_IDS, { name: 'patch', type: 'string', desc: 'Object of fields to set' }] },
  { id: 'setVolume', label: 'Volume', desc: 'Set clip volume (0..1)', hasUI: true, params: [CLIP_IDS, { name: 'volume', type: 'number', desc: '0..1' }] },
  { id: 'setAudioFade', label: 'Audio fade', desc: 'Set audio fade in/out (s)', hasUI: true, params: [CLIP_IDS, { name: 'audioFadeIn', type: 'number', optional: true, desc: 'Seconds' }, { name: 'audioFadeOut', type: 'number', optional: true, desc: 'Seconds' }] },
  { id: 'setOpacity', label: 'Opacity', desc: 'Set clip opacity (0..1)', hasUI: true, params: [CLIP_IDS, { name: 'opacity', type: 'number', desc: '0..1' }] },
  { id: 'setFade', label: 'Fade', desc: 'Set visual fade in/out (s)', hasUI: true, params: [CLIP_IDS, { name: 'fadeIn', type: 'number', optional: true, desc: 'Seconds' }, { name: 'fadeOut', type: 'number', optional: true, desc: 'Seconds' }] },
  { id: 'setSpeed', label: 'Speed', desc: 'Set playback speed multiplier', hasUI: true, params: [CLIP_IDS, { name: 'speed', type: 'number', desc: 'e.g. 0.5, 2' }] },
  { id: 'setColor', label: 'Color grade', desc: 'Set brightness/contrast/saturation', hasUI: true, params: [CLIP_IDS, { name: 'brightness', type: 'number', optional: true, desc: '1 = normal' }, { name: 'contrast', type: 'number', optional: true, desc: '1 = normal' }, { name: 'saturation', type: 'number', optional: true, desc: '1 = normal' }] },
  { id: 'setScale', label: 'Scale', desc: 'Set clip scale', hasUI: true, params: [CLIP_IDS, { name: 'scale', type: 'number', desc: '1 = fit' }] },
  { id: 'setRotation', label: 'Rotate to', desc: 'Set absolute rotation (deg)', hasUI: true, params: [CLIP_IDS, { name: 'degrees', type: 'number', desc: 'Degrees' }] },
  { id: 'rotateBy', label: 'Rotate by', desc: 'Rotate by a delta (deg)', hasUI: true, params: [CLIP_IDS, { name: 'degrees', type: 'number', desc: 'Delta degrees' }] },
  { id: 'nudge', label: 'Nudge', desc: 'Offset clip position (fractions of canvas)', hasUI: false, params: [CLIP_IDS, { name: 'dx', type: 'number', desc: '-1..1' }, { name: 'dy', type: 'number', desc: '-1..1' }] },
  { id: 'resetTransform', label: 'Reset transform', desc: 'Clear scale/position/rotation', hasUI: true, params: [CLIP_IDS] },
  { id: 'flipH', label: 'Flip horizontal', desc: 'Mirror left-right', hasUI: true, params: [CLIP_IDS] },
  { id: 'flipV', label: 'Flip vertical', desc: 'Mirror top-bottom', hasUI: true, params: [CLIP_IDS] },
  { id: 'setCrop', label: 'Crop', desc: 'Reveal only a sub-rectangle (mask). Null clears it.', hasUI: true, params: [CLIP_IDS, { name: 'crop', type: 'rect', optional: true, desc: '{x,y,w,h} 0..1 or null' }] },
  { id: 'link', label: 'Link', desc: 'Group clips so they move together', hasUI: true, params: [CLIP_IDS] },
  { id: 'unlink', label: 'Unlink', desc: 'Break a link group', hasUI: true, params: [CLIP_IDS] },
  { id: 'addText', label: 'Add text', desc: 'Add a text clip', hasUI: true, params: [{ name: 'at', type: 'number', desc: 'Timeline seconds' }, { name: 'text', type: 'string', optional: true, desc: 'Text content' }, { name: 'trackId', type: 'trackId', optional: true, desc: 'Text track' }] },
  { id: 'addTrack', label: 'Add track', desc: 'Add a video/audio/text track', hasUI: true, params: [{ name: 'kind', type: 'string', desc: 'video|audio|text' }] },
  { id: 'removeTrack', label: 'Remove track', desc: 'Delete a track and its clips', hasUI: true, params: [{ name: 'trackId', type: 'trackId', desc: 'Track id' }] },
  { id: 'moveTrack', label: 'Move track', desc: 'Reorder a track up/down', hasUI: true, params: [{ name: 'trackId', type: 'trackId', desc: 'Track id' }, { name: 'dir', type: 'number', desc: '-1 up, 1 down' }] },
  { id: 'setTrackHidden', label: 'Show/hide track', desc: 'Toggle a track visible/audible', hasUI: true, params: [{ name: 'trackId', type: 'trackId', desc: 'Track id' }, { name: 'hidden', type: 'boolean', desc: 'true = hidden' }] },
  { id: 'setTrackLocked', label: 'Lock track', desc: 'Toggle a track locked', hasUI: true, params: [{ name: 'trackId', type: 'trackId', desc: 'Track id' }, { name: 'locked', type: 'boolean', desc: 'true = locked' }] },
]

// Compact JSON schema string to hand to an AI as its tool catalogue.
export function commandSchemaJSON(): string {
  return JSON.stringify(
    COMMAND_REGISTRY.map((c) => ({
      type: c.id,
      description: c.desc,
      params: c.params.map((p) => ({ name: p.name, type: p.type, optional: !!p.optional, desc: p.desc })),
    })),
    null,
    2,
  )
}

File: native/src/commands.rs
//! Editor action layer — Rust port of `editor/commands.ts`. A single typed
//! vocabulary of edits; both the toolbar/inspector and (later) the AI agent go
//! through `apply_command`, so every button is one code path and trivially
//! undoable.

use crate::model::{ChromaKey, ColorGrade, Keyframe, MaskRect, TimelineData, Transform};
use crate::ops;

#[derive(Debug, Clone)]
pub enum Command {
    // structure
    Split { clip_id: String, at: f64 },
    Delete { clip_ids: Vec<String> },
    RippleDelete { clip_ids: Vec<String> },
    Duplicate { clip_ids: Vec<String> },
    MoveClip { clip_id: String, start: f64, track_id: Option<String> },
    TrimStart { clip_id: String, at: f64 },
    TrimEnd { clip_id: String, at: f64, source_max: Option<f64> },
    // layer / fades / audio
    SetOpacity { clip_ids: Vec<String>, opacity: f64 },
    SetFade { clip_ids: Vec<String>, fade_in: Option<f64>, fade_out: Option<f64> },
    SetVolume { clip_ids: Vec<String>, volume: f64 },
    SetAudioFade { clip_ids: Vec<String>, fade_in: Option<f64>, fade_out: Option<f64> },
    SetSpeed { clip_ids: Vec<String>, speed: f64 },
    // color
    SetColor { clip_ids: Vec<String>, brightness: Option<f64>, contrast: Option<f64>, saturation: Option<f64> },
    // transform
    SetScale { clip_ids: Vec<String>, scale: f64 },
    SetTransform { clip_ids: Vec<String>, transform: Transform },
    SetRotation { clip_ids: Vec<String>, degrees: f64 },
    RotateBy { clip_ids: Vec<String>, degrees: f64 },
    Nudge { clip_ids: Vec<String>, dx: f64, dy: f64 },
    ResetTransform { clip_ids: Vec<String> },
    FlipH { clip_ids: Vec<String> },
    FlipV { clip_ids: Vec<String> },
    SetFlip { clip_ids: Vec<String>, flip_h: Option<bool>, flip_v: Option<bool> },
    // crop / mask / chroma
    SetCrop { clip_ids: Vec<String>, crop: Option<MaskRect> },
    SetMask { clip_ids: Vec<String>, mask: Option<MaskRect> },
    SetChroma { clip_ids: Vec<String>, chroma: Option<ChromaKey> },
    // text
    SetText { clip_id: String, text: String },
    AddText { at: f64, track_id: Option<String>, text: Option<String>, x: Option<f64>, y: Option<f64> },
    // keyframes
    AddKeyframeAtPlayhead { clip_ids: Vec<String>, playhead: f64 },
    RemoveKeyframe { clip_id: String, t: f64 },
    // linking
    Link { clip_ids: Vec<String> },
    Unlink { clip_ids: Vec<String> },
    // tracks
    AddTrack { kind: String },
    RemoveTrack { track_id: String },
    MoveTrack { track_id: String, dir: i64 },
    SetTrackHidden { track_id: String, hidden: bool },
    SetTrackLocked { track_id: String, locked: bool },
}

fn merge_transform(el_transform: Option<Transform>, apply: impl FnOnce(&mut Transform)) -> Transform {
    let mut t = el_transform.unwrap_or_default();
    apply(&mut t);
    t
}

pub fn apply_command(data: &mut TimelineData, cmd: &Command) {
    match cmd {
        Command::Split { clip_id, at } => ops::split_clip(data, clip_id, *at),
        Command::Delete { clip_ids } => ops::delete_clips(data, clip_ids),
        Command::RippleDelete { clip_ids } => ops::ripple_delete(data, clip_ids),
        Command::Duplicate { clip_ids } => ops::duplicate_clips(data, clip_ids),
        Command::MoveClip { clip_id, start, track_id } => {
            ops::move_clip_group(data, clip_id, *start, track_id.as_deref())
        }
        Command::TrimStart { clip_id, at } => ops::trim_start(data, clip_id, *at),
        Command::TrimEnd { clip_id, at, source_max } => ops::trim_end(data, clip_id, *at, *source_max),

        Command::SetOpacity { clip_ids, opacity } => {
            for id in clip_ids {
                ops::update_clip(data, id, |e| e.opacity = Some(opacity.clamp(0.0, 1.0)));
            }
        }
        Command::SetFade { clip_ids, fade_in, fade_out } => {
            for id in clip_ids {
                ops::update_clip(data, id, |e| {
                    if let Some(v) = fade_in { e.fade_in = Some(v.max(0.0)); }
                    if let Some(v) = fade_out { e.fade_out = Some(v.max(0.0)); }
                });
            }
        }
        Command::SetVolume { clip_ids, volume } => {
            for id in clip_ids {
                ops::update_clip(data, id, |e| e.volume = Some(volume.clamp(0.0, 1.0)));
            }
        }
        Command::SetAudioFade { clip_ids, fade_in, fade_out } => {
            for id in clip_ids {
                ops::update_clip(data, id, |e| {
                    if let Some(v) = fade_in { e.audio_fade_in = Some(v.max(0.0)); }
                    if let Some(v) = fade_out { e.audio_fade_out = Some(v.max(0.0)); }
                });
            }
        }
        Command::SetSpeed { clip_ids, speed } => {
            for id in clip_ids {
                ops::update_clip(data, id, |e| e.speed = Some(speed.max(0.1)));
            }
        }
        Command::SetColor { clip_ids, brightness, contrast, saturation } => {
            for id in clip_ids {
                ops::update_clip(data, id, |e| {
                    let cur = e.color.unwrap_or_default();
                    e.color = Some(ColorGrade {
                        brightness: brightness.unwrap_or(cur.brightness),
                        contrast: contrast.unwrap_or(cur.contrast),
                        saturation: saturation.unwrap_or(cur.saturation),
                    });
                });
            }
        }
        Command::SetScale { clip_ids, scale } => {
            for id in clip_ids {
                ops::update_clip(data, id, |e| {
                    e.transform = Some(merge_transform(e.transform, |t| t.scale = scale.max(0.05)));
                });
            }
        }
        Command::SetTransform { clip_ids, transform } => {
            for id in clip_ids {
                ops::update_clip(data, id, |e| e.transform = Some(*transform));
            }
        }
        Command::SetRotation { clip_ids, degrees } => {
            for id in clip_ids {
                ops::update_clip(data, id, |e| {
                    e.transform = Some(merge_transform(e.transform, |t| t.rotation = *degrees));
                });
            }
        }
        Command::RotateBy { clip_ids, degrees } => {
            for id in clip_ids {
                ops::update_clip(data, id, |e| {
                    let cur = e.transform.unwrap_or_default().rotation;
                    e.transform = Some(merge_transform(e.transform, |t| t.rotation = cur + *degrees));
                });
            }
        }
        Command::Nudge { clip_ids, dx, dy } => {
            for id in clip_ids {
                ops::update_clip(data, id, |e| {
                    let cur = e.transform.unwrap_or_default();
                    e.transform = Some(merge_transform(e.transform, |t| {
                        t.x = cur.x + *dx;
                        t.y = cur.y + *dy;
                    }));
                });
            }
        }
        Command::ResetTransform { clip_ids } => {
            for id in clip_ids {
                ops::update_clip(data, id, |e| e.transform = Some(Transform::default()));
            }
        }
        Command::FlipH { clip_ids } => {
            for id in clip_ids {
                ops::update_clip(data, id, |e| e.flip_h = Some(!e.flip_h.unwrap_or(false)));
            }
        }
        Command::FlipV { clip_ids } => {
            for id in clip_ids {
                ops::update_clip(data, id, |e| e.flip_v = Some(!e.flip_v.unwrap_or(false)));
            }
        }
        Command::SetFlip { clip_ids, flip_h, flip_v } => {
            for id in clip_ids {
                ops::update_clip(data, id, |e| {
                    if let Some(v) = flip_h { e.flip_h = Some(*v); }
                    if let Some(v) = flip_v { e.flip_v = Some(*v); }
                });
            }
        }
        Command::SetCrop { clip_ids, crop } => {
            for id in clip_ids {
                ops::update_clip(data, id, |e| e.crop = *crop);
            }
        }
        Command::SetMask { clip_ids, mask } => {
            for id in clip_ids {
                ops::update_clip(data, id, |e| e.mask = *mask);
            }
        }
        Command::SetChroma { clip_ids, chroma } => {
            for id in clip_ids {
                ops::update_clip(data, id, |e| e.chroma = chroma.clone());
            }
        }
        Command::SetText { clip_id, text } => {
            ops::update_clip(data, clip_id, |e| e.text = Some(text.clone()));
        }
        Command::AddText { at, track_id, text, x, y } => {
            let tid = track_id
                .clone()
                .or_else(|| data.tracks.iter().find(|t| t.kind == "text").map(|t| t.id.clone()));
            if let Some(tid) = tid {
                let mut clip = ops::make_text_clip(text.as_deref().unwrap_or("New text"), *at);
                if x.is_some() || y.is_some() {
                    clip.transform = Some(Transform {
                        x: x.unwrap_or(0.0),
                        y: y.unwrap_or(0.0),
                        ..Transform::default()
                    });
                }
                ops::add_clip(data, &tid, clip);
            }
        }
        Command::AddKeyframeAtPlayhead { clip_ids, playhead } => {
            let ids = clip_ids.clone();
            for id in ids {
                let Some(el) = data.get(&id).cloned() else { continue };
                if el.is_audio() {
                    continue;
                }
                let dur = el.duration();
                let clt = (playhead - el.timeline_start).clamp(0.0, dur);
                let p = ops::resolve_props(&el, clt, dur);
                ops::add_keyframe(
                    data,
                    &id,
                    Keyframe {
                        t: clt,
                        opacity: Some(p.opacity),
                        scale: Some(p.scale),
                        x: Some(p.x),
                        y: Some(p.y),
                        rotation: Some(p.rotation),
                    },
                );
            }
        }
        Command::RemoveKeyframe { clip_id, t } => ops::remove_keyframe(data, clip_id, *t),
        Command::Link { clip_ids } => ops::link_clips(data, clip_ids),
        Command::Unlink { clip_ids } => ops::unlink_clips(data, clip_ids),
        Command::AddTrack { kind } => ops::add_track(data, kind, 0),
        Command::RemoveTrack { track_id } => ops::remove_track(data, track_id),
        Command::MoveTrack { track_id, dir } => ops::move_track(data, track_id, *dir),
        Command::SetTrackHidden { track_id, hidden } => {
            ops::set_track_flags(data, track_id, Some(*hidden), None)
        }
        Command::SetTrackLocked { track_id, locked } => {
            ops::set_track_flags(data, track_id, None, Some(*locked))
        }
    }
}
